### PR TITLE
router: ability to use query parameters in HTTP hash policy

### PR DESCRIPTION
**Pull request by zuercher**
Risk Level: low
Testing: unit tests
Doc Changes: n/a
Release Notes: updated

Signed-off-by: Stephan Zuercher <zuercher@gmail.com>

### CODE REVIEW

File: api/envoy/api/v2/route/route.proto [hunk 1]
```diff
@@ -604,6 +605,13 @@ message RouteAction {
       bool source_ip = 1;
     }
 
+    message QueryParameter {
+      // The name of the URL query parameter that will be used to obtain the hash
+      // key. If the parameter is not present, no hash will be produced. Query
+      // parameter names are case-sensitive.
+      string parameter_name = 1 [(validate.rules).string = {min_bytes: 1}];
```
Comment (htuch) [inline api/envoy/api/v2/route/route.proto]: From a @envoyproxy/api-shepherds  perspective, I think `name` is shorter and unambiguous here, so suggest switching to that.

File: test/common/router/config_impl_test.cc [hunk 2]
```diff
@@ -2197,6 +2197,26 @@ TEST_F(RouterMatcherHashPolicyTest, HashIpv6DifferentAddresses) {
...
+  }
+  {
+    Http::TestHeaderMapImpl headers = genHeaders("www.lyft.com", "/foo?param=xyz", "GET");
+    Router::RouteConstSharedPtr route = config().route(headers, 0);
+    EXPECT_TRUE(route->routeEntry()->hashPolicy()->generateHash(nullptr, headers, add_cookie_nop_));
+  }
+  {
+    Http::TestHeaderMapImpl headers = genHeaders("www.lyft.com", "/bar?param=xyz", "GET");
+    Router::RouteConstSharedPtr route = config().route(headers, 0);
+    EXPECT_EQ(nullptr, route->routeEntry()->hashPolicy());
```
Comment (snowp) [inline test/common/router/config_impl_test.cc]: this expectation seems inconsistent with the other two (checking against bool converted ptr vs equality)

Comment (zuercher) [reply test/common/router/config_impl_test.cc]: ✅

**Comment by zuercher** [comment]
@snowp if you’re around, could you give this a quick look

**Comment by zuercher** [comment]
@htuch and @snop unfortunately I accidentally closed it and had to merge master again. 🤦‍♂️

**Pull request merged**